File: packages/tracelet_platform_interface/lib/src/algorithms/location_processor.dart
import 'adaptive_sampling_engine.dart';
import 'geo_utils.dart';

/// Result of processing a location through [LocationProcessor].
///
/// When [accepted] is `true`, the location passed all filters and should be
/// delivered to the user. When `false`, the location was rejected and the
/// [reason] and optional [errorMessage] describe why.
class LocationProcessorResult {
  const LocationProcessorResult._({
    required this.accepted,
    this.effectiveSpeed = 0,
    this.odometerDelta = 0,
    this.distance = 0,
    this.reason,
    this.errorMessage,
    this.isError = false,
  });

  /// The location passed all filters.
  factory LocationProcessorResult.accept({
    required double effectiveSpeed,
    required double odometerDelta,
    required double distance,
  }) => LocationProcessorResult._(
    accepted: true,
    effectiveSpeed: effectiveSpeed,
    odometerDelta: odometerDelta,
    distance: distance,
  );

  /// The location was silently filtered (not delivered to the user).
  factory LocationProcessorResult.filtered(String reason) =>
      LocationProcessorResult._(accepted: false, reason: reason);

  /// The location was filtered and an error event should be dispatched.
  factory LocationProcessorResult.error(String reason, String message) =>
      LocationProcessorResult._(
        accepted: false,
        reason: reason,
        errorMessage: message,
        isError: true,
      );

  /// Whether the location was accepted by all filters.
  final bool accepted;

  /// Computed effective speed in m/s (platform speed or distance/time).
  final double effectiveSpeed;

  /// Distance (meters) to add to the odometer for this location.
  final double odometerDelta;

  /// Distance (meters) from the previous accepted location.
  final double distance;

  /// Filter name that rejected the location (e.g. `DISTANCE_FILTER`).
  final String? reason;

  /// Human-readable error message for `discard`-policy rejections.
  final String? errorMessage;

  /// Whether the rejection should dispatch an error event to the user.
  final bool isError;

  @override
  String toString() => accepted
      ? 'LocationProcessorResult.accept(speed=$effectiveSpeed, '
            'odometerDelta=$odometerDelta, distance=$distance)'
      : 'LocationProcessorResult.${isError ? "error" : "filtered"}'
            '(reason=$reason${errorMessage != null ? ', msg=$errorMessage' : ''})';
}

/// Pure-Dart location filtering engine.
///
/// Replaces the filtering logic previously duplicated in native Kotlin
/// (`LocationEngine.onLocationReceived`) and Swift
/// (`LocationEngine.didUpdateLocations`).
///
/// **Filters applied (in order):**
/// 1. **Elasticity** — dynamically scales `distanceFilter` based on speed.
/// 2. **Distance filter** — rejects locations closer than `effectiveDistance`.
/// 3. **Accuracy filter** — rejects locations exceeding
///    `trackingAccuracyThreshold` according to the `filterPolicy`.
/// 4. **Speed filter** — rejects impossible jumps exceeding `maxImpliedSpeed`.
/// 5. **Odometer gating** — only adds distance to the odometer when accuracy
///    is within `odometerAccuracyThreshold`.
///
/// **This is a pure Dart implementation** — no native code required. It runs
/// identically on Android, iOS, web, macOS, Linux, and Windows.
///
/// ```dart
/// final processor = LocationProcessor(distanceFilter: 10);
/// final result = processor.process(
///   latitude: 37.422,
///   longitude: -122.084,
///   accuracy: 16.0,
///   speed: 1.5,
///   timestampMs: DateTime.now().millisecondsSinceEpoch,
/// );
/// if (result.accepted) {
///   print('Accepted, speed=${result.effectiveSpeed}');
/// }
/// ```
class LocationProcessor {
  LocationProcessor({
    this.distanceFilter = 10.0,
    this.disableElasticity = false,
    this.elasticityMultiplier = 1.0,
    this.enableAdaptiveMode = false,
    this.trackingAccuracyThreshold = 0,
    this.filterPolicy = 0,
    this.maxImpliedSpeed = 0,
    this.odometerAccuracyThreshold = 0,
    this.rejectMockLocations = false,
    this.mockDetectionLevel = 1,
    this.enableSparseUpdates = false,
    this.sparseDistanceThreshold = 50.0,
    this.sparseMaxIdleSeconds = 300,
  });

  /// Cached adaptive sampling engine — avoids re-creation per GPS fix.
  late final AdaptiveSamplingEngine _adaptiveEngine = AdaptiveSamplingEngine(
    baseDistanceFilter: distanceFilter,
    elasticityMultiplier: elasticityMultiplier,
  );

  /// Base distance filter in meters.
  final double distanceFilter;

  /// When `true`, elasticity scaling is disabled and [distanceFilter] is used
  /// as-is regardless of speed.
  final bool disableElasticity;

  /// Multiplier applied to the elasticity-scaled distance.
  /// Values < 1 make the filter more aggressive, > 1 more relaxed.
  final double elasticityMultiplier;

  /// When `true`, the adaptive sampling engine computes the distance filter
  /// using activity type, battery level, and charging state instead of
  /// simple speed-based elasticity. Supply [AdaptiveContext] to [process]
  /// for full benefit. When `false` (default), classic elasticity is used.
  final bool enableAdaptiveMode;

  /// Maximum acceptable GPS accuracy in meters. Locations above this
  /// threshold are handled according to [filterPolicy].
  /// `0` disables the accuracy filter.
  final int trackingAccuracyThreshold;

  /// How to handle locations that exceed [trackingAccuracyThreshold].
  ///
  /// - `0` (adjust): Skip the location but don't fire an error event.
  ///   Only applies when there is a previous location; the very first
  ///   location is always accepted.
  /// - `1` (ignore): Silently drop the location.
  /// - `2` (discard): Drop the location and fire an error event.
  final int filterPolicy;

  /// Maximum plausible speed in m/s. If the implied speed between two
  /// consecutive locations exceeds this, the location is rejected.
  /// `0` disables the speed filter.
  final int maxImpliedSpeed;

  /// Maximum GPS accuracy (meters) for the distance to count towards the
  /// odometer. Locations with lower accuracy still pass through but don't
  /// add to the odometer. `0` disables the threshold (all count).
  final int odometerAccuracyThreshold;

  /// When `true`, locations flagged as mock/spoofed (`isMock == true`) are
  /// automatically rejected. The rejection behavior follows [filterPolicy]:
  /// - `0` / `1`: silent drop.
  /// - `2`: drop and fire an error event.
  final bool rejectMockLocations;

  /// Mock detection level.
  ///
  /// - `0` (disabled): No mock detection.
  /// - `1` (basic): Only platform-flag-based detection.
  /// - `2` (heuristic): Platform flags + native heuristics (satellite count,
  ///   elapsed realtime drift) + Dart-side timestamp monotonicity check.
  final int mockDetectionLevel;

  /// Enable sparse updates (intelligent deduplication).
  ///
  /// When `true`, locations that haven't moved beyond
  /// [sparseDistanceThreshold] from the last *recorded* location are
  /// silently dropped — unless [sparseMaxIdleSeconds] has elapsed.
  final bool enableSparseUpdates;

  /// Minimum distance (meters) to move before recording a new location
  /// when sparse updates are enabled.
  final double sparseDistanceThreshold;

  /// Maximum idle time (seconds) before forcing a recording even if the
  /// device hasn't moved beyond [sparseDistanceThreshold].
  /// `0` disables idle timeout (only movement triggers recording).
  final int sparseMaxIdleSeconds;

  // ─────────────────────────────────────────────────────────────────────────
  // Internal state
  // ─────────────────────────────────────────────────────────────────────────

  double? _lastLatitude;
  double? _lastLongitude;
  int _lastTimestampMs = 0;

  /// Last recorded location for sparse deduplication.
  double? _sparseLastLat;
  double? _sparseLastLng;
  int _sparseLastTimestampMs = 0;

  /// Last computed effective speed in m/s.
  double lastEffectiveSpeed = 0;

  /// Whether [process] has been called at least once and accepted a location.
  bool get hasLastLocation => _lastLatitude != null;

  /// Process a new location and return the filter decision.
  ///
  /// - [latitude], [longitude]: GPS coordinates in degrees.
  /// - [accuracy]: Horizontal accuracy in meters.
  /// - [speed]: Platform-reported speed in m/s. Pass `-1` or `0` if
  ///   unavailable — a fallback speed will be computed from distance/time.
  /// - [timestampMs]: Location timestamp in milliseconds since epoch.
  /// - [adaptiveContext]: Optional context for adaptive sampling (battery,
  ///   activity, charging). Only used when [enableAdaptiveMode] is `true`.
  LocationProcessorResult process({
    required double latitude,
    required double longitude,
    required double accuracy,
    required double speed,
    required int timestampMs,
    bool isMock = false,
    AdaptiveContext? adaptiveContext,
  }) {
    // ── Mock location filter ──────────────────────────────────────────────
    if (rejectMockLocations && isMock) {
      if (filterPolicy == 2) {
        return LocationProcessorResult.error(
          'MOCK_LOCATION',
          'Location rejected: flagged as mock/spoofed by the platform',
        );
      }
      return LocationProcessorResult.filtered('MOCK_LOCATION');
    }

    // ── Timestamp monotonicity (heuristic level) ──────────────────────────
    // When mockDetectionLevel >= 2 and we have a previous timestamp, reject
    // locations whose timestamp goes backwards. Real GPS hardware never
    // produces decreasing timestamps; replayed / injected locations can.
    if (mockDetectionLevel >= 2 &&
        rejectMockLocations &&
        _lastTimestampMs > 0 &&
        timestampMs < _lastTimestampMs) {
      if (filterPolicy == 2) {
        return LocationProcessorResult.error(
          'MOCK_LOCATION_TIMESTAMP',
          'Location rejected: timestamp $timestampMs is before previous '
              '$_lastTimestampMs (non-monotonic — possible replay attack)',
        );
      }
      return LocationProcessorResult.filtered('MOCK_LOCATION_TIMESTAMP');
    }

    // ── Distance & speed computation ──────────────────────────────────────
    double distance = 0;
    double timeDelta = 0;

    if (_lastLatitude != null && _lastLongitude != null) {
      distance = GeoUtils.haversine(
        _lastLatitude!,
        _lastLongitude!,
        latitude,
        longitude,
      );
      timeDelta = (timestampMs - _lastTimestampMs) / 1000.0;
    }

    final computedSpeed = (distance > 0 && timeDelta > 0)
        ? distance / timeDelta
        : 0.0;
    final effectiveSpeed = (speed > 0) ? speed : computedSpeed;

    // ── Elasticity / Adaptive: scale distanceFilter ──────────────────────
    double effectiveDistance = distanceFilter;
    if (enableAdaptiveMode) {
      final ctx = adaptiveContext ?? AdaptiveContext(speed: effectiveSpeed);
      // Ensure the engine sees the computed speed even if the caller didn't
      // set it in the context.
      final enriched = AdaptiveContext(
        batteryLevel: ctx.batteryLevel,
        isCharging: ctx.isCharging,
        activityType: ctx.activityType,
        activityConfidence: ctx.activityConfidence,
        speed: ctx.speed > 0 ? ctx.speed : effectiveSpeed,
      );
      final result = _adaptiveEngine.compute(enriched);
      effectiveDistance = result.effectiveDistanceFilter;
    } else if (!disableElasticity && effectiveSpeed > 0) {
      final multiplier = elasticityMultiplier < 0.1
          ? 0.1
          : elasticityMultiplier;
      final speedFactor = (effectiveSpeed / 10.0).clamp(1.0, 10.0);
      effectiveDistance = distanceFilter * speedFactor * multiplier;
    }

    // ── Distance filter ───────────────────────────────────────────────────
    if (_lastLatitude != null && distance < effectiveDistance) {
      return LocationProcessorResult.filtered('DISTANCE_FILTER');
    }

    // ── Accuracy filter ───────────────────────────────────────────────────
    if (trackingAccuracyThreshold > 0 && accuracy > trackingAccuracyThreshold) {
      switch (filterPolicy) {
        case 2: // discard — fire error event
          return LocationProcessorResult.error(
            'ACCURACY_FILTER',
            'Location accuracy ${accuracy}m exceeds '
                'threshold ${trackingAccuracyThreshold}m',
          );
        case 1: // ignore — silent drop
          return LocationProcessorResult.filtered('ACCURACY_FILTER');
        default: // adjust — skip if we already have a reference location
          if (_lastLatitude != null) {
            return LocationProcessorResult.filtered('ACCURACY_FILTER');
          }
      }
    }

    // ── Speed filter ──────────────────────────────────────────────────────
    if (maxImpliedSpeed > 0 && _lastLatitude != null && timeDelta > 0) {
      final impliedSpeed = distance / timeDelta;
      if (impliedSpeed > maxImpliedSpeed) {
        if (filterPolicy == 2) {
          return LocationProcessorResult.error(
            'SPEED_FILTER',
            'Implied speed ${impliedSpeed.toStringAsFixed(1)}m/s exceeds '
                'max ${maxImpliedSpeed}m/s',
          );
        }
        return LocationProcessorResult.filtered('SPEED_FILTER');
      }
    }

    // ── Odometer gating ───────────────────────────────────────────────────
    double odometerDelta = 0;
    if (odometerAccuracyThreshold <= 0 ||
        accuracy <= odometerAccuracyThreshold) {
      odometerDelta = distance;
    }

    // ── Sparse deduplication ──────────────────────────────────────────────
    if (enableSparseUpdates) {
      if (_sparseLastLat != null && _sparseLastLng != null) {
        final sparseDist = GeoUtils.haversine(
          _sparseLastLat!,
          _sparseLastLng!,
          latitude,
          longitude,
        );
        final sparseElapsed = (timestampMs - _sparseLastTimestampMs) / 1000.0;

        final withinDistance = sparseDist < sparseDistanceThreshold;
        final withinTime =
            sparseMaxIdleSeconds == 0 || sparseElapsed < sparseMaxIdleSeconds;

        if (withinDistance && withinTime) {
          // Update the base distance/speed state but don't record.
          _lastLatitude = latitude;
          _lastLongitude = longitude;
          _lastTimestampMs = timestampMs;
          lastEffectiveSpeed = effectiveSpeed;
          return LocationProcessorResult.filtered('SPARSE_FILTER');
        }
      }
      _sparseLastLat = latitude;
      _sparseLastLng = longitude;
      _sparseLastTimestampMs = timestampMs;
    }

    // ── Accept ────────────────────────────────────────────────────────────
    _lastLatitude = latitude;
    _lastLongitude = longitude;
    _lastTimestampMs = timestampMs;
    lastEffectiveSpeed = effectiveSpeed;

    return LocationProcessorResult.accept(
      effectiveSpeed: effectiveSpeed,
      odometerDelta: odometerDelta,
      distance: distance,
    );
  }

  /// Reset all internal state. Call when tracking restarts.
  void reset() {
    _lastLatitude = null;
    _lastLongitude = null;
    _lastTimestampMs = 0;
    lastEffectiveSpeed = 0;
    _sparseLastLat = null;
    _sparseLastLng = null;
    _sparseLastTimestampMs = 0;
  }

  /// Create a new [LocationProcessor] with updated configuration.
  ///
  /// Preserves internal state (last position, timestamp) so filtering
  /// continuity is maintained across config changes.
  LocationProcessor copyWith({
    double? distanceFilter,
    bool? disableElasticity,
    double? elasticityMultiplier,
    bool? enableAdaptiveMode,
    int? trackingAccuracyThreshold,
    int? filterPolicy,
    int? maxImpliedSpeed,
    int? odometerAccuracyThreshold,
    bool? rejectMockLocations,
    int? mockDetectionLevel,
    bool? enableSparseUpdates,
    double? sparseDistanceThreshold,
    int? sparseMaxIdleSeconds,
  }) {
    final copy = LocationProcessor(
      distanceFilter: distanceFilter ?? this.distanceFilter,
      disableElasticity: disableElasticity ?? this.disableElasticity,
      elasticityMultiplier: elasticityMultiplier ?? this.elasticityMultiplier,
      enableAdaptiveMode: enableAdaptiveMode ?? this.enableAdaptiveMode,
      trackingAccuracyThreshold:
          trackingAccuracyThreshold ?? this.trackingAccuracyThreshold,
      filterPolicy: filterPolicy ?? this.filterPolicy,
      maxImpliedSpeed: maxImpliedSpeed ?? this.maxImpliedSpeed,
      odometerAccuracyThreshold:
          odometerAccuracyThreshold ?? this.odometerAccuracyThreshold,
      rejectMockLocations: rejectMockLocations ?? this.rejectMockLocations,
      mockDetectionLevel: mockDetectionLevel ?? this.mockDetectionLevel,
      enableSparseUpdates: enableSparseUpdates ?? this.enableSparseUpdates,
      sparseDistanceThreshold:
          sparseDistanceThreshold ?? this.sparseDistanceThreshold,
      sparseMaxIdleSeconds: sparseMaxIdleSeconds ?? this.sparseMaxIdleSeconds,
    );
    // Preserve internal state.
    copy._lastLatitude = _lastLatitude;
    copy._lastLongitude = _lastLongitude;
    copy._lastTimestampMs = _lastTimestampMs;
    copy.lastEffectiveSpeed = lastEffectiveSpeed;
    copy._sparseLastLat = _sparseLastLat;
    copy._sparseLastLng = _sparseLastLng;
    copy._sparseLastTimestampMs = _sparseLastTimestampMs;
    return copy;
  }

  /// Copy internal tracking state (odometer position, timestamps, speed,
  /// sparse-update bookkeeping) from this processor into [target].
  ///
  /// Use this when replacing a processor with updated config parameters
  /// while preserving continuity of distance/speed calculations.
  void transferStateTo(LocationProcessor target) {
    target._lastLatitude = _lastLatitude;
    target._lastLongitude = _lastLongitude;
    target._lastTimestampMs = _lastTimestampMs;
    target.lastEffectiveSpeed = lastEffectiveSpeed;
    target._sparseLastLat = _sparseLastLat;
    target._sparseLastLng = _sparseLastLng;
    target._sparseLastTimestampMs = _sparseLastTimestampMs;
  }
}

File: sdk/rust-core/src/algorithms/location_processor.rs
use crate::algorithms::geo_utils::haversine;
use std::sync::Mutex;

/// Represents the interpreted physical activity state of the device.
#[derive(uniffi::Enum, Clone, Copy, PartialEq)]
pub enum ActivityType {
    Still,
    Walking,
    Running,
    OnFoot,
    InVehicle,
    OnBicycle,
    Unknown,
}

/// Confidence level of the associated physical activity.
#[derive(uniffi::Enum, Clone, Copy, PartialEq)]
pub enum ActivityConfidence {
    Low,
    Medium,
    High,
}

/// Environmental context used to adapt location sampling frequency.
#[derive(uniffi::Record)]
pub struct AdaptiveContext {
    pub battery_level: f64,
    pub is_charging: bool,
    pub activity_type: ActivityType,
    pub activity_confidence: ActivityConfidence,
    pub speed: f64,
}

impl Default for AdaptiveContext {
    fn default() -> Self {
        Self {
            battery_level: -1.0,
            is_charging: false,
            activity_type: ActivityType::Unknown,
            activity_confidence: ActivityConfidence::Low,
            speed: 0.0,
        }
    }
}

/// The primary factor driving the current adaptive sampling rate.
#[derive(uniffi::Enum, Clone, Copy, PartialEq)]
pub enum AdaptiveSource {
    Activity,
    Speed,
    Static,
}

/// Results from evaluating the current context to determine the optimal sampling parameters.
#[derive(uniffi::Record)]
pub struct AdaptiveSamplingResult {
    pub effective_distance_filter: f64,
    pub base_distance_filter: f64,
    pub activity_factor: f64,
    pub battery_factor: f64,
    pub speed_factor: f64,
    pub source: AdaptiveSource,
}

/// Core logic engine for dynamically adjusting distance filters based on context.
#[derive(uniffi::Object)]
pub struct AdaptiveSamplingEngine {
    base_distance_filter: f64,
    elasticity_multiplier: f64,
}

#[uniffi::export]
impl AdaptiveSamplingEngine {
    #[uniffi::constructor]
    pub fn new(base_distance_filter: f64, elasticity_multiplier: f64) -> Self {
        Self {
            base_distance_filter,
            elasticity_multiplier,
        }
    }

    pub fn compute(&self, context: AdaptiveContext) -> AdaptiveSamplingResult {
        let mut activity_factor = 1.0;
        let mut speed_factor = 1.0;
        let mut source = AdaptiveSource::Static;

        let use_activity = context.activity_type != ActivityType::Unknown
            && context.activity_confidence != ActivityConfidence::Low;

        if use_activity {
            let activity_distance = Self::activity_distance(context.activity_type);
            activity_factor = activity_distance / self.base_distance_filter;
            source = AdaptiveSource::Activity;
        } else if context.speed > 0.0 {
            let mult = self.elasticity_multiplier.max(0.1);
            speed_factor = (context.speed / 10.0).clamp(1.0, 10.0) * mult;
            source = AdaptiveSource::Speed;
        }

        let batt_factor = Self::battery_factor(context.battery_level, context.is_charging);

        let effective = match source {
            AdaptiveSource::Activity => self.base_distance_filter * activity_factor * batt_factor,
            AdaptiveSource::Speed => self.base_distance_filter * speed_factor * batt_factor,
            AdaptiveSource::Static => self.base_distance_filter * batt_factor,
        };

        AdaptiveSamplingResult {
            effective_distance_filter: effective,
            base_distance_filter: self.base_distance_filter,
            activity_factor,
            battery_factor: batt_factor,
            speed_factor,
            source,
        }
    }
}

impl AdaptiveSamplingEngine {
    const DISTANCE_STILL: f64 = 500.0;
    const DISTANCE_WALKING: f64 = 50.0;
    const DISTANCE_RUNNING: f64 = 30.0;
    const DISTANCE_BICYCLE: f64 = 25.0;
    const DISTANCE_VEHICLE: f64 = 10.0;

    const BATTERY_HIGH_THRESHOLD: f64 = 0.50;
    const BATTERY_MEDIUM_THRESHOLD: f64 = 0.20;
    const BATTERY_LOW_THRESHOLD: f64 = 0.10;

    const BATTERY_MEDIUM_FACTOR: f64 = 1.5;
    const BATTERY_LOW_FACTOR: f64 = 2.5;
    const BATTERY_CRITICAL_FACTOR: f64 = 5.0;

    fn activity_distance(activity: ActivityType) -> f64 {
        match activity {
            ActivityType::Still => Self::DISTANCE_STILL,
            ActivityType::Walking | ActivityType::OnFoot => Self::DISTANCE_WALKING,
            ActivityType::Running => Self::DISTANCE_RUNNING,
            ActivityType::OnBicycle => Self::DISTANCE_BICYCLE,
            ActivityType::InVehicle => Self::DISTANCE_VEHICLE,
            ActivityType::Unknown => 10.0,
        }
    }

    fn battery_factor(battery_level: f64, is_charging: bool) -> f64 {
        if is_charging || battery_level < 0.0 {
            return 1.0;
        }
        if battery_level < Self::BATTERY_LOW_THRESHOLD {
            return Self::BATTERY_CRITICAL_FACTOR;
        }
        if battery_level < Self::BATTERY_MEDIUM_THRESHOLD {
            return Self::BATTERY_LOW_FACTOR;
        }
        if battery_level < Self::BATTERY_HIGH_THRESHOLD {
            return Self::BATTERY_MEDIUM_FACTOR;
        }
        1.0
    }
}

/// Represents the outcome of filtering and processing a single location update.
#[derive(uniffi::Record)]
pub struct LocationProcessorResult {
    pub accepted: bool,
    pub effective_speed: f64,
    pub odometer_delta: f64,
    pub distance: f64,
    pub reason: Option<String>,
    pub error_message: Option<String>,
    pub is_error: bool,
}

impl LocationProcessorResult {
    fn accept(effective_speed: f64, odometer_delta: f64, distance: f64) -> Self {
        Self {
            accepted: true,
            effective_speed,
            odometer_delta,
            distance,
            reason: None,
            error_message: None,
            is_error: false,
        }
    }

    fn filtered(reason: &str) -> Self {
        Self {
            accepted: false,
            effective_speed: 0.0,
            odometer_delta: 0.0,
            distance: 0.0,
            reason: Some(reason.to_string()),
            error_message: None,
            is_error: false,
        }
    }

    fn error(reason: &str, message: &str) -> Self {
        Self {
            accepted: false,
            effective_speed: 0.0,
            odometer_delta: 0.0,
            distance: 0.0,
            reason: Some(reason.to_string()),
            error_message: Some(message.to_string()),
            is_error: true,
        }
    }
}

struct LocationProcessorState {
    last_latitude: Option<f64>,
    last_longitude: Option<f64>,
    last_timestamp_ms: i64,
    sparse_last_lat: Option<f64>,
    sparse_last_lng: Option<f64>,
    sparse_last_timestamp_ms: i64,
    last_effective_speed: f64,
}

/// Core location processing engine that handles filtering out inaccurate or redundant points.
#[derive(uniffi::Object)]
pub struct LocationProcessor {
    distance_filter: f64,
    disable_elasticity: bool,
    elasticity_multiplier: f64,
    enable_adaptive_mode: bool,
    tracking_accuracy_threshold: i32,
    filter_policy: i32,
    max_implied_speed: i32,
    odometer_accuracy_threshold: i32,
    reject_mock_locations: bool,
    mock_detection_level: i32,
    enable_sparse_updates: bool,
    sparse_distance_threshold: f64,
    sparse_max_idle_seconds: i32,
    state: Mutex<LocationProcessorState>,
}

#[uniffi::export]
impl LocationProcessor {
    #[uniffi::constructor]
    pub fn new(
        distance_filter: f64,
        disable_elasticity: bool,
        elasticity_multiplier: f64,
        enable_adaptive_mode: bool,
        tracking_accuracy_threshold: i32,
        filter_policy: i32,
        max_implied_speed: i32,
        odometer_accuracy_threshold: i32,
        reject_mock_locations: bool,
        mock_detection_level: i32,
        enable_sparse_updates: bool,
        sparse_distance_threshold: f64,
        sparse_max_idle_seconds: i32,
    ) -> Self {
        Self {
            distance_filter,
            disable_elasticity,
            elasticity_multiplier,
            enable_adaptive_mode,
            tracking_accuracy_threshold,
            filter_policy,
            max_implied_speed,
            odometer_accuracy_threshold,
            reject_mock_locations,
            mock_detection_level,
            enable_sparse_updates,
            sparse_distance_threshold,
            sparse_max_idle_seconds,
            state: Mutex::new(LocationProcessorState {
                last_latitude: None,
                last_longitude: None,
                last_timestamp_ms: 0,
                sparse_last_lat: None,
                sparse_last_lng: None,
                sparse_last_timestamp_ms: 0,
                last_effective_speed: 0.0,
            }),
        }
    }

    pub fn last_effective_speed(&self) -> f64 {
        self.state.lock().unwrap().last_effective_speed
    }

    pub fn has_last_location(&self) -> bool {
        self.state.lock().unwrap().last_latitude.is_some()
    }

    pub fn process(
        &self,
        latitude: f64,
        longitude: f64,
        accuracy: f64,
        speed: f64,
        timestamp_ms: i64,
        is_mock: bool,
        adaptive_context: Option<AdaptiveContext>,
    ) -> LocationProcessorResult {
        let mut state = self.state.lock().unwrap();

        if self.reject_mock_locations && is_mock {
            return if self.filter_policy == 2 {
                LocationProcessorResult::error(
                    "MOCK_LOCATION",
                    "Location rejected: flagged as mock/spoofed by the platform",
                )
            } else {
                LocationProcessorResult::filtered("MOCK_LOCATION")
            };
        }

        if self.mock_detection_level >= 2
            && self.reject_mock_locations
            && state.last_timestamp_ms > 0
            && timestamp_ms < state.last_timestamp_ms
        {
            return if self.filter_policy == 2 {
                LocationProcessorResult::error(
                    "MOCK_LOCATION_TIMESTAMP",
                    &format!(
                        "Location rejected: timestamp {} is before previous {} (non-monotonic)",
                        timestamp_ms, state.last_timestamp_ms
                    ),
                )
            } else {
                LocationProcessorResult::filtered("MOCK_LOCATION_TIMESTAMP")
            };
        }

        let mut distance = 0.0;
        let mut time_delta = 0.0;

        if let (Some(prev_lat), Some(prev_lng)) = (state.last_latitude, state.last_longitude) {
            distance = haversine(prev_lat, prev_lng, latitude, longitude);
            time_delta = (timestamp_ms - state.last_timestamp_ms) as f64 / 1000.0;
        }

        let computed_speed = if distance > 0.0 && time_delta > 0.0 {
            distance / time_delta
        } else {
            0.0
        };
        let effective_speed = if speed > 0.0 { speed } else { computed_speed };

        let mut effective_distance = self.distance_filter;
        if self.enable_adaptive_mode {
            let mut ctx = adaptive_context.unwrap_or_default();
            if ctx.speed <= 0.0 {
                ctx.speed = effective_speed;
            }
            let engine = AdaptiveSamplingEngine::new(self.distance_filter, self.elasticity_multiplier);
            effective_distance = engine.compute(ctx).effective_distance_filter;
        } else if !self.disable_elasticity && effective_speed > 0.0 {
            let multiplier = self.elasticity_multiplier.max(0.1);
            let speed_factor = (effective_speed / 10.0).clamp(1.0, 10.0);
            effective_distance = self.distance_filter * speed_factor * multiplier;
        }

        if state.last_latitude.is_some() && distance < effective_distance {
            return LocationProcessorResult::filtered("DISTANCE_FILTER");
        }

        if self.tracking_accuracy_threshold > 0 && accuracy > self.tracking_accuracy_threshold as f64 {
            match self.filter_policy {
                2 => {
                    return LocationProcessorResult::error(
                        "ACCURACY_FILTER",
                        &format!(
                            "Location accuracy {}m exceeds threshold {}m",
                            accuracy, self.tracking_accuracy_threshold
                        ),
                    )
                }
                1 => return LocationProcessorResult::filtered("ACCURACY_FILTER"),
                _ => {
                    if state.last_latitude.is_some() {
                        return LocationProcessorResult::filtered("ACCURACY_FILTER");
                    }
                }
            }
        }

        if self.max_implied_speed > 0 && state.last_latitude.is_some() && time_delta > 0.0 {
            let implied_speed = distance / time_delta;
            if implied_speed > self.max_implied_speed as f64 {
                return if self.filter_policy == 2 {
                    LocationProcessorResult::error(
                        "SPEED_FILTER",
                        &format!(
                            "Implied speed {:.1}m/s exceeds max {}m/s",
                            implied_speed, self.max_implied_speed
                        ),
                    )
                } else {
                    LocationProcessorResult::filtered("SPEED_FILTER")
                };
            }
        }

        let odometer_delta = if self.odometer_accuracy_threshold <= 0
            || accuracy <= self.odometer_accuracy_threshold as f64
        {
            distance
        } else {
            0.0
        };

        if self.enable_sparse_updates {
            if let (Some(s_lat), Some(s_lng)) = (state.sparse_last_lat, state.sparse_last_lng) {
                let sparse_dist = haversine(s_lat, s_lng, latitude, longitude);
                let sparse_elapsed = (timestamp_ms - state.sparse_last_timestamp_ms) as f64 / 1000.0;

                let within_distance = sparse_dist < self.sparse_distance_threshold;
                let within_time = self.sparse_max_idle_seconds == 0
                    || sparse_elapsed < self.sparse_max_idle_seconds as f64;

                if within_distance && within_time {
                    state.last_latitude = Some(latitude);
                    state.last_longitude = Some(longitude);
                    state.last_timestamp_ms = timestamp_ms;
                    state.last_effective_speed = effective_speed;
                    return LocationProcessorResult::filtered("SPARSE_FILTER");
                }
            }
            state.sparse_last_lat = Some(latitude);
            state.sparse_last_lng = Some(longitude);
            state.sparse_last_timestamp_ms = timestamp_ms;
        }

        state.last_latitude = Some(latitude);
        state.last_longitude = Some(longitude);
        state.last_timestamp_ms = timestamp_ms;
        state.last_effective_speed = effective_speed;

        LocationProcessorResult::accept(effective_speed, odometer_delta, distance)
    }

    pub fn reset(&self) {
        let mut state = self.state.lock().unwrap();
        state.last_latitude = None;
        state.last_longitude = None;
        state.last_timestamp_ms = 0;
        state.last_effective_speed = 0.0;
        state.sparse_last_lat = None;
        state.sparse_last_lng = None;
        state.sparse_last_timestamp_ms = 0;
    }
}
